### PR TITLE
Using hackage security

### PR DESCRIPTION
**Pull request by edsko**
This PR makes a few improvements to the `hackage-mirror` client, and adds instructions to the `README` for setting up the security infrastructure.

### CODE REVIEW

**Comment by edsko** [comment]
(Travis fails because the server is started without the required security files in place.)

**Comment by ocharles** [comment]
There is a merge conflict here, fixed in https://github.com/ocharles/hackage-server (`master`).

**Comment by edsko** [comment]
I will create a new PR that fixes the tests, and then rebase this stuff on that.

**Comment by edsko** [comment]
I've rebased this PR on `master`. However, **THIS IS WIP**. I've now added example keys to the server so that people can get a Hackage server going without first having to generate keys etc. However, this now all relies on the latest `hackage-security` library, which isn't released yet. I'd release it, but it in turn relies on the latest `tar` library, which @dcoutts hasn't released yet and he's on holidays. The tests will also still fail on this branch, even if it did compile, for an annoying reason which I will fix as well.

In short: working on this, will indicate when it's ready to be merged.

**Comment by edsko** [comment]
Ok guys. This is done! Tests pass! They won't yet on Travis because it relies on yet-unreleased versions of `tar` and `hackage-security`, but once they are released we can restart the Travis job for this PR and it should be green :punch: :boom: Happy new year :)

An overview of the commits in this PR:

* https://github.com/edsko/hackage-server/commit/ffa205e1642d622dd98e338b2531b900f118f40d is #459 
* https://github.com/edsko/hackage-server/commit/6c20397785ee6a86c81da0c0c3bf3ba25a176739 adds example keys to the repo, along with corresponding TUF files in `datafiles/TUF`. This means the server can run on Travis and also makes it easier for people to get their own instances up and running.
* https://github.com/edsko/hackage-server/commit/bc2bd8b226654a0432362e1c0ff4a292803ffd9a just updates the code base to use `hackage-security-0.5`.
* https://github.com/edsko/hackage-server/commit/fdbc895ba2a10eec9e35034b9bf97ef8869bff1d , https://github.com/edsko/hackage-server/commit/1c44e50897c8e05e75e21568e16acdbf1afcb5f4 , https://github.com/edsko/hackage-server/commit/8dba414381e55711671c8a971efc715b84c45f01 and https://github.com/edsko/hackage-server/commit/c2d6fe2b43ee437117077543e348100f1cba63d2 are changes to the mirror client. These are already in use on the server, where we were using a branch even though it had not been merged yet with HEAD here. (The commits are mostly unchanged, minor changes here and there to allow for the newer `hackage-security` library.)
* https://github.com/edsko/hackage-server/commit/16bf0517fe63d63ab1be0fec4485d37ebcaf5fa7 just adds instructions to the README on how to set up the TUF infrastructure, and explains the situation with the example keys.
* https://github.com/edsko/hackage-server/commit/b507a0974c895f0daca5e14bcd6b34d6ffb885bd replaces the use of the (slow) `SHA` package with the (much faster) `cryptohash` package, just like we have done in the `hackage-security`. Although the `Digest SHA256` type from the `SHA` package was used as part of the server state, this change does *not* need a state migration because the the binary instance is fully backwards compatible (this is explained in detail in the commit message).
* https://github.com/edsko/hackage-server/commit/24feca8307507126e1a33c58204b753a38e4d110 then adds a new field `securityLastUpdate` into the `SecurityState`. This *does* require a state migration, but a simple one: it just sets the `securityLastUpdate `field to 'Nothing'. This is a very small commit so that this migration can easily be audited separately.
* The remainder of the commits (https://github.com/edsko/hackage-server/commit/67ab723c67377896ad64ea580d09cc2869ffaaec , https://github.com/edsko/hackage-server/commit/62478273a7724266ebf8a60e0630dab7c7aaf8f9 and https://github.com/haskell/hackage-server/commit/6f47c2b3412db3ee0a803e2c0fb4d640cb992816) then finally make use of this new information to avoid unnecessarily updating the snapshot and timestamp. This means that simply restarting the server, or importing a backup, does not automatically create a new snapshot/timestamp, and that means that the server backup roundtrips now pass.

**Comment by edsko** [comment]
I have a snapshot of the Hackage server state on my local machine (updated yesterday). I intended to run the server (using the same commit as on the live server right now) on this, then update to my PR and test the migration. However, I cannot run the server on my snapshot:

```
hackage-server: Guessing public URI as http://server.local:8080
(you can override with the --base-uri= flag)
hackage-server: state/db/PackagesState/checkpoints-0000000013.log: invalid
argument
```

I don't know what this means. I tried asking the server to create a checkpoint and updated my local snapshot after that, but that didn't help either. It's a little worrying that this is not working, makes me wonder about the integrity of the server state on the live server. (I tried running my local server on a Hackage snapshot in the past and it worked back then.)

**Comment by edsko** [comment]
Note that I have not modified the generation of the source distribution; I don't know if we want to make sure that the example TUF files as part of that; possibly we do. I've opened a separate ticket #465 about this.

**Comment by dcoutts** [comment]
Merged. 765ec003ac9fe4d3084d6487ead93818ac4367b7
